linux_container.go

```diff
@@ -121,76 +315,275 @@ func (c *linuxContainer) updateStateFile() error {
+		fmt.Sprintf("_LIBCONTAINER_INITPID=%d", ppid),
+		fmt.Sprintf("_LIBCONTAINER_USERNS=1"))
+	err = command.Start()
+	child.Close()
+	if err != nil {
+		return err
+	}
+	s, err := command.Process.Wait()
+	if err != nil {
+		return err
+	}
+	if !s.Success() {
+		return &exec.ExitError{s}
+	}
+	decoder := json.NewDecoder(parent)
+	var pid *pid
+	if err := decoder.Decode(&pid); err != nil {
+		return err
+	}
+	p, err := os.FindProcess(pid.Pid)
+	if err != nil {
+		return err
+	}
+	terminate := func(terr error) error {
+		// TODO: log the errors for kill and wait
+		p.Kill()
+		p.Wait()
+		return terr
+	}
+	// send the state to the container's init process then shutdown writes for the parent
+	if err := json.NewEncoder(parent).Encode(process); err != nil {
+		return terminate(err)
+	}
+	// shutdown writes for the parent side of the pipe
+	if err := syscall.Shutdown(int(parent.Fd()), syscall.SHUT_WR); err != nil {
+		return terminate(err)
+	}
+	// wait for the child process to fully complete and receive an error message
+	// if one was encoutered
+	var ierr *initError
+	if err := decoder.Decode(&ierr); err != nil && err != io.EOF {
+		return terminate(err)
+	}
+	if ierr != nil {
+		return ierr
+	}
+	s, err = p.Wait()
+	if err != nil {
+		return err
+	}
+	if !s.Success() {
+		return &exec.ExitError{s}
 	}
-
-	os.RemoveAll(c.root)
 	return nil
 }
 
-func (c *linuxContainer) Pause() error {
-	return c.cgroupManager.Freeze(cgroups.Frozen)
+type pid struct {
+	Pid int `json:"Pid"`
 }
 
-func (c *linuxContainer) Resume() error {
-	return c.cgroupManager.Freeze(cgroups.Thawed)
+// Finalize entering into a container and execute a specified command
+func InitIn(pipe *os.File) (err error) {
+	defer func() {
+		// if we have an error during the initialization of the container's init then send it back to the
+		// parent process in the form of an initError.
+		if err != nil {
+			// ensure that any data sent from the parent is consumed so it doesn't
+			// receive ECONNRESET when the child writes to the pipe.
+			ioutil.ReadAll(pipe)
+			if err := json.NewEncoder(pipe).Encode(initError{
+				Message: err.Error(),
+			}); err != nil {
+				panic(err)
+			}
+		}
+		// ensure that this pipe is always closed
+		pipe.Close()
+	}()
+	decoder := json.NewDecoder(pipe)
+	var config *configs.Config
+	if err := decoder.Decode(&config); err != nil {
+		return err
+	}
+	var process *processArgs
+	if err := decoder.Decode(&process); err != nil {
+		return err
+	}
+	if err := finalizeSetns(config); err != nil {
+		return err
+	}
+	if err := system.Execv(process.Args[0], process.Args[0:], config.Env); err != nil {
+		return err
+	}
+	panic("unreachable")
 }
 
-func (c *linuxContainer) Signal(pid, signal int) error {
-	glog.Infof("sending signal %d to pid %d", signal, pid)
-	panic("not implemented")
-}
+// finalize expects that the setns calls have been setup and that is has joined an
+// existing namespace
+func finalizeSetns(container *configs.Config) error {
```

Comment:
Ya, i was going to take another pass after this one is merged in refineing more func and method names to make sense.